### PR TITLE
[FLINK-1718] Adds sparse matrix and sparse vector

### PR DESCRIPTION
**Pull request by tillrohrmann**
Adds a sparse matrix abstraction using CSC data representation. Adds a sparse vector abstraction using CSC data representation.

Adds convenience functions to cast a Flink matrix/vector into a Breeze matrix/vector and vice versa.

### CODE REVIEW

File: flink-staging/flink-ml/pom.xml [hunk 1]
```diff
@@ -46,9 +46,9 @@
 		</dependency>
 
 		<dependency>
-			<groupId>com.github.fommil.netlib</groupId>
```
Comment (hsaputra) [inline flink-staging/flink-ml/pom.xml]: Just curious, why was this deleted?

Comment (tillrohrmann) [inline flink-staging/flink-ml/pom.xml]: Because Breeze depends on it and thus it is pulled transitivly.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/DenseVector.scala [hunk 2]
```diff
@@ -67,7 +63,25 @@ case class DenseVector(val values: Array[Double]) extends Vector {
...
+    DenseVector(data.clone())
+  }
+
+  /** Updates the element at the given index with the provided value
+    *
+    * @param index
+    * @param value
+    */
+  override def update(index: Int, value: Double): Unit = {
+    require(0 <= index && index < data.length, s"Index $index is out of bounds " +
```
Comment (rmetzger) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/DenseVector.scala]: Might be inefficient.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/SparseMatrix.scala [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+      result.append(" " * (valueFieldWidth - valueStr.length) + valueStr)
+      result.append("\n")
+    }
+
+    result.toString
+  }
+
+  private def locate(row: Int, col: Int): Int = {
+    require(0 <= row && row < numRows, s"Row $row is out of bounds [0, $numRows).")
+    require(0 <= col && col < numCols, s"Col $col is out of bounds [0, $numCols).")
```
Comment (rmetzger) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/SparseMatrix.scala]: Many ops in this class seem to use this method. I'm not sure how expensive the string interpolation is.

**Comment by rmetzger** [comment]
Looks good to merge.
There are some performance improvements possible.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/SparseMatrix.scala (hunk 3 above)
Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/SparseMatrix.scala]: That's right. I'll probably remove the checks and expect the user to know what he's doing.

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/math/SparseMatrix.scala]: Well but actually the string interpolations do not costs anything, because the message parameter is called by-name. Meaning that it is only evaluated in case of a false requirement.

**Comment by tillrohrmann** [comment]
I combined some of the sanity checks of the matrix/vector accessors.

**Comment by tillrohrmann** [comment]
I rebased the code on the current master and unified some of the tests. When Travis confirms that everything works and if there are no objections, I'll merge the PR.

**Comment by tillrohrmann** [comment]
FYI: Travis failed only for the last profile because the tests didn't start for some reason. Travis passed for my own repository, though.
